README
    LIBNATSPEC
	The Nation Specifity Library
	Vitaly Lipatov <lav@etersoft.ru>

== About libnatspec ===

libnatspec is a collection of function for requiest various
charsets and locales for host system and for other system.
The libnatspec allows applications to be written that are
locale and system independent. Libnatspec is tries to use
user locale information if possible.
It contains table with relations between locale name and encoding
of other system. 
It tested in various OS (with GCC installed):
 + ALT Linux 2.4 Master (x86)
 + Debian GNU/Linux 3.0 (Alpha)
 - Solaris SunOS 5.9 (Sparc/x86) (missed libiconv during linking)
 ? Slackware (x86)
 + Mandrake (x86)
 + RedHat (x86)
 ? ASPLinux (x86)
 - FreeBSD 4.10 (x86) (missed popt)
It is not tested in other Unix-like system yet.

It has been tested with various Linux programs/servers:
 + mount
 + mkisofs / isoinfo
 + smbmount
 + submount (subfs)
 - smbd/nmbd/swat/smbclient

This distribution contains library, header files and binding
to other languages as python (--with-python option for configure)

Please if you discover some bug or misfunction of the natspec library
on your system, email me about it.

=== Installation ===

Installation instructions are found in INSTALL.
For RPM-based system you can use libnatspec.spec as an example.

=== More Information ===

LIBNATSPEC is at sourceforge.net called as natspec project.
libnatspec also has a web page, at http://etersoft.ru/natspec
or http://natspec.sourceforge.net

New versions of the package can be found on the ftp site
ftp://ftp.altlinux.ru/pub/people/lav/natspec or the http site
http://etersoft.ru/download/natspec

libnatspec may be freely distibuted and modified in accordance with
the GNU Lesser General Public License. Examples in the libnatspec
distribution can fall under different licenses;
see the example/LICENSE file and individual files for details.

The person for contact is Vitaly Lipatov <lav@etersoft.ru>

=== Examples of use ===

==== Convert files from other system to our one ====

// Get the filename encoding in your system
const char *fileenc = natspec_get_filename_encoding("");

// Get the charset in other system WIN
const char *win_cs = natspec_get_charset_by_locale(NATSPEC_WINCS,"");

// Do conversion from foreign to local filename encoding
const char * unix_filename = natspec_convert(win_filename,win_cs, fileenc);


==== Using in .c files ====

#ifdef HAVE_NATSPEC
#include <natspec.h>
#endif
...
#ifdef HAVE_NATSPEC
	some with natspec using
#endif

==== Using in configure.in files ====
1. Add line with AM_PATH_NATSPEC in configure.in
it makes NATSPEC_LIBS, NATSPEC_CFLAGS variables for Makefiles
and defines HAVE_NATSPEC in config.h
2. Use option --with-natspec for configure

==== Links ====
http://www.sensi.org/~alec/locale/ About locale

07.05.2005


NEWS
libnatspec-0.2.4: 15-Feb-2008:
 - natspec_convert returns original string if failed (fix bug #14301)
 - small bugfixes, clean code

libnatspec-0.2.3: 02-Jan-2008:
 - add patches for unzip, pmount
 - fix build

libnatspec-0.2.3-beta: 13-Mar-2006:
 - add natspec_iconv, natspec_iconv_open, natspec-iconv_close
 - add transliterate param to natspec_convert (set to 0 for old behaviour)
 - obsolete natspec_convert_with_translit (use natspec_convert with last param set to 1)
 - add --system-locale in natspec
 - add patch for o3read
 - add patch for catpkt from Igor Zubkov <icesik@mail.ru>
 - check G_FILENAME_ENCODING for @locale value
 - add patch for dict client, isoinfo from cdrtools
 
libnatspec-0.2.2: 22-Jul-2005:
 - fix use static buffer natspec_get_codepage (change type of returned value to int)
 - review code and some cleanup
 - introduce natspec_locale_is_utf8 function
 - do not use LANGUAGE env variable
 - build API documentation

libnatspec-0.2.1: 09-Apr-2005:
 - fix CP1125 order
 - fix locale string parsing
 - add print current charset in natspec util (in info mode)

libnatspec-0.2.0: 17-Mar-2005:
 - rename get_user_locale() to get_current_locale()
 - discard from toupper/tolower/isalnum/strcasecmp using (it was locale depended)
 - fix DOS charset for uk_UA
 - introduce samba/loadparams (for server autotuning) - does not work yet

libnatspec-0.1.9: 03-Mar-2005:
 - add jfs support
 - update submount patch (to be simple)
 - update samba patch (add default charsets for params)
 - fixes for compilation in various OS
 - natspec: print overall information by default, print ASCII for unknown system
 - do not use setlocale anywhere
 - add --with-i18n-file option for set file with LANG variable at your system
 - add --with-python option to configure
 - improved get_filesystem_encoding
 - humble_charset do toupper now and breaks at ':' symbol
 - repack_locale do tolower for lang and toupper for country string

libnatspec-0.1.0: 27-Feb-2005:
 - fixed locale using order, add some fallbacks for unusual situation
 - update docs

libnatspec-0.0.9: 27-Feb-2005:
 - rename enrich_fs_options to get_enriched_fs_options with incompatible change args
 - update mount/submount, sdcv example patches
 - add example path for samba (smbmount/cifsmount)
 - add auto assumed local encoding in natspec_convert
 - add convert_with_translit with transliteration
   (from links project) if conversion is not possible
 - add natspec.m4 for autotools, it makes natspec configure is easy

libnatspec-0.0.4: 24-Feb-2005:
 - add python binding
 - code cleanup
 - cleanup comments in header file
 - add README with short description (in English)
 - add README-ru.html with description (in Russian)
 
libnatspec-0.0.3: 23-Feb-2005:
 - fix bug with unix charset (was due misunderstanging wine behaviour)
 - add examples of patch for mount, submount, mkisofs, isoinfo, stardict-sdcv
 
libnatspec-0.0.2: 22-Feb-2005:
 - add get_charset function
 - add convert for lossy conversion between encoding
 - add smbfs support in enrich_fs_options

libnatspec-0.0.1: 20-Feb-2005:
 - first public release after born
 - support for mount filesystem (enrich_fs_options)
 - system filename encoding detection (get_filename_encoding)
 - get system locale from some file (/etc/sysconfig/i18n) (get_system_locale)
 - retrieve charset/codepage for another OS by system/current
   locale/charset (get_charset_by_locale)
 - command line util for using features in shells (natspec)


src/natspec.c
/*
    src/natspec.c
	
	command line util for access to libnatspec functions

    NATSPEC library
    Copyright (c) 2005 Etersoft
    Copyright (c) 2005 Vitaly Lipatov <lav@etersoft.ru>

    $Id: natspec.c,v 1.22 2005/09/03 18:59:28 vitlav Exp $

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation, version 2.1
    of the License.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
    USA.
*/

#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <assert.h>
#include <locale.h>

#include "natspec_internal.h"
#include "natspechi_internal.h"

#ifdef HAVE_LIBPOPT
#  include <popt.h>
#endif

#include <locale.h>
#ifdef HAVE_LANGINFO_H
# include <langinfo.h>
#endif


char *charset_type;
char *locale, *transliterate;
int version, verbose, fsenc, fcodepage, nls, flag_help, info;
int flag_locale, utf8, country_id, get_system_locale;

#if defined HAVE_LIBPOPT
poptContext context = NULL;
struct poptOption options[] =
{
    {"codepage", 'c', POPT_ARG_NONE,  &fcodepage, 1,
     "print codepage", ""},
    {"fsenc", 'f', POPT_ARG_NONE,  &fsenc, 1,
     "print filesystem encoding", ""},
	{"utf8", 0, POPT_ARG_NONE,  &utf8, 1,
     "return 0 if locale is utfed", ""},
	{"country-id", 0, POPT_ARG_NONE,  &country_id, 1,
     "print country ID (international numbering)", ""},
	{"system-locale", 0, POPT_ARG_NONE,  &get_system_locale, 1,
     "print system locale", ""},
    {"info", 'i', POPT_ARG_NONE,  &info, 0,
     "print overall encoding/charset info for your system", ""},
    {"transl", 'a', POPT_ARG_STRING,  &transliterate, 1,
     "transliterate args from ENC to current encoding and print it", ""},
    {"version", 'V', POPT_ARG_NONE, &version, 1,
     "display version and exit", NULL },
    {"verbose", 'v', POPT_ARG_NONE, &verbose, 0,
     "verbose output", NULL },
    {"charset", 's', POPT_ARG_STRING,  &charset_type, 0,
     "print charset for op. system: unix, win, dos, mac (case insensivity)", ""},
    {"help", 'h', POPT_ARG_NONE, &flag_help, 1, "Show this help message" },
    {(char *) NULL, '\0', 0, NULL, 0}
};
#endif

static const char *get_charset(char *charset_type)
{
	/* FIXME: strcasecmp is locale depends */
	const char *charset;
	int type;
	assert (charset_type);
	if (!strcasecmp(charset_type,"WIN"))
		type = NATSPEC_WINCS;
	else if (!strcasecmp(charset_type,"DOS"))
		type = NATSPEC_DOSCS;
	else if (!strcasecmp(charset_type,"MAC"))
		type = NATSPEC_MACCS;
	else if (!strcasecmp(charset_type,"UNIX"))
		type = NATSPEC_UNIXCS;
	else
	{
		if (verbose) printf("We do not know %s type of encoding\n",charset_type);
		printf("ASCII\n");
		exit(1);
	}
	charset = natspec_get_charset_by_locale(type, locale);
	if (verbose)
		printf("Charset of '%s' system: %s\n", charset_type, charset);
	else
		printf("%s\n",charset);
	return charset;
}


int main(int argc, const char** argv)
{
	int is_utf8;
#if defined HAVE_LIBPOPT
	int rc = 0;
	poptContext poptCtx;
	poptCtx = poptGetContext("natspec", argc, (const char **)argv, options, 0);
	poptSetOtherOptionHelp(poptCtx, "[OPTION...]");
	while (rc >= 0) {
		if((rc = poptGetNextOpt(poptCtx)) < -1) {
			fprintf(stderr, "Error on option %s: %s.\nRun '%s --help' to see a full list of available command line options.\n",
				poptBadOption(poptCtx, 0),
				poptStrerror(rc),
				argv[0]);
			exit(1);
		}
	}
/*
  { int flag_help=1;
  if (argv) {
      while (poptPeekArg(poptCtx))
		flag_help=0;
		//locale = (char *)poptGetArg(poptCtx);
      poptFreeContext(poptCtx);
	}
*/	
    if(flag_help)
	{
	poptPrintHelp(poptCtx, stdout, 0);
    exit(0);
	}

#else
	fprintf(stderr, "Compiled without popt. Exit\n");
	exit(1);
#endif
	if (argc == 1)
		info = 1;
	if (version || info)
	{
		printf("%s, compiled %s\n",PACKAGE_STRING,__DATE__);
		if (!info) exit(0);
	}
	if (transliterate)
	{
		char *str = NULL;
		if (argv) {
			while (poptPeekArg(poptCtx))
				str = (char *)poptGetArg(poptCtx);
			poptFreeContext(poptCtx);
		}
		if ( str != NULL)
		{
			setlocale(LC_ALL,"");
			str = natspec_convert_with_translit(str, "", transliterate);
			puts(str);
		}
		exit(0);
	}
	if (info)
	{
		printf(" === Overall information ===\n");
		verbose = 1;
	}
	locale = natspec_get_current_locale();
	if (flag_locale || info)
	{
		if (verbose) printf("Current locale: ");
		printf("%s",locale);
		if (verbose) puts("");
	}
	if (get_system_locale || verbose)
	{
		if (verbose) printf("System locale: ");
		printf("%s", natspec_get_system_locale());
		if (verbose) puts("");
		if (!info) exit(0);
	}
	if (fsenc || info)
	{
		const char *buf;
		if (verbose) printf("Filename encoding in iconv/nls form: ");
		buf = natspec_get_filename_encoding("");
		printf("%s",buf);
		if (verbose) puts("");
		if (!info) exit(0);
	}
	if (country_id || info)
	{
		int id;
		if (verbose) printf("Country ID (for DOS): ");
//		id = natspec_get_country_id();
		printf("%d", id);
		if (verbose) puts("");
		if (!info) exit(0);
	}
	if (fcodepage || info)
	{
		int cp;
		if (verbose) printf("Codepage of DOS: ");
		cp = natspec_get_codepage_by_charset(
			natspec_get_charset_by_locale(NATSPEC_DOSCS, locale));
		printf("%d", cp);
		if (verbose) puts("");
		if (!info) exit(0);
	}
	if (info)
	{
		char *types[]={"UNIX","WIN","DOS","MAC"};
		int i;
		for (i=0; i<sizeof(types)/sizeof(char*);i++)
			get_charset(types[i]);
		setlocale(LC_ALL,"");
		printf("After setlocale:\n");
#ifdef HAVE_LANGINFO_H
		printf("\tnl_langinfo(CODESET): %s\n",nl_langinfo(CODESET));
#endif
		printf("\tnatspec_get_charset: %s\n",natspec_get_charset());
	}
	if (charset_type)
		get_charset(charset_type);
	is_utf8 = natspec_locale_is_utf8("");
	if (verbose)
		printf("Current locale is%sin UTF8 encoding\n",(is_utf8 ? " " : " NOT "));
	if (utf8 && !info)
		return is_utf8 ? 0 : 1;
	return 0;
}
